### ISSUE TITLE
Lookup.run not working on Mac OS Big Sur

### ISSUE THREAD

**Issue by dea18** [dnsjava/dnsjava#194]
I have recently upgraded my Macbook to Big Sur and now the Lookup.run function is no longer working. The DNS lookup seems to timeout and then the function returns a null. This had worked perfectly fine for me when my Macbook was running Mojave.

I am using dnsjava v3.0.2 however I have also tried v3.3.1. I am also using Java SDK 1.8.

Please can I get some assistance.

**Comment by ibauersachs** [comment]
Please provide more details, I cannot possibly know what exactly you're doing. A good start would be logs at debug or trace level from dnsjava. Also, please make sure the detected nameservers are actually reachable, see #185 for an example.

**Comment by dea18** [comment]
I normally invoke the Lookup.run function from within my Gatling performance test code which makes it tricky to debug.

I have written this scala object to help me debug the issue:
`import org.xbill.DNS._

object dns {
  def main(args: Array[String]): Unit = {
    try {
      val lookup = new Lookup("some_url", Type.A)
      val records = lookup.run().toList
      println(records)
    } catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }
  }
}`

This is the stack trace that was caught:
`java.lang.NullPointerException
	at scala.collection.mutable.ArrayOps$ofRef$.length$extension(ArrayOps.scala:204)
	at scala.collection.mutable.ArrayOps$ofRef.length(ArrayOps.scala:204)
	at scala.collection.IndexedSeqOptimized.toList(IndexedSeqOptimized.scala:285)
	at scala.collection.IndexedSeqOptimized.toList$(IndexedSeqOptimized.scala:284)
	at scala.collection.mutable.ArrayOps$ofRef.toList(ArrayOps.scala:198)
	at common.dns$.main(dns.scala:10)
	at common.dns.main(dns.scala)`

It is worth noting that the URL that I am trying to resolve, resolves when I use dig from my terminal and my Gatling tests work absolutely fine when I run them locally in a Docker container. It seems to only be a problem when I run the code straight on my local machine (without Docker).

Does this help or do you need some more info?

**Comment by nresare** [comment]
If Lookup.run() returns null, you should call Lookup.getErrorString() on it to see what failed.

**Comment by dea18** [comment]
That returns that the look up timed out which is what I suspected from the beginning but the question is why is it timing out as a dig works fine and the dnsjava Lookup works fine when run on my local machine from a Docker container?

**Comment by dea18** [comment]
Something that I have just discovered is that if I clean and then compile, then the above mentioned code works fine every time I run it. As soon as I try to do multiple look ups, the first look up succeeds and the second one times out. Here is the code that I am running for the multi look ups:
```
import org.xbill.DNS._

object dns {
  def main(args: Array[String]): Unit = {
    try {
      val lookup = new Lookup("url_1", Type.A)
      val records = lookup.run()
      if (records == null) {
        println(lookup.getErrorString)
      }
      val list = records.toList
      println(list)
    } catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }
    try {
      val lookup = new Lookup("url_2", Type.A)
      val records = lookup.run()
      if (records == null) {
        println(lookup.getErrorString)
      }
      val list = records.toList
      println(list)
    } catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }
  }
}
```

The above is more representative of what is being done in my Gatling test. My Gatling test loops through a list of URLs to be looked up.

**Comment by ibauersachs** [comment]
As I mentioned in my first reply, please attach the debug (or trace) log output from dnsjava. It uses slf4j-api, so you can attach basically any logging framework. For your example, even slf4j-simple works. If you're using sbt:

`libraryDependencies += "org.slf4j" % "slf4j-simple" % "1.7.30"`

Also pass `-Dorg.slf4j.simpleLogger.defaultLogLevel=debug` (or trace) as a JVM argument.

**Comment by dea18** [comment]
Thanks for the pointers. Here are the trace logs:

```
[main] DEBUG org.xbill.DNS.config.JndiContextResolverConfigProvider$InnerJndiContextResolverConfigProvider - JNDI class: javax.naming.directory.DirContext
[main] DEBUG org.xbill.DNS.config.ResolvConfResolverConfigProvider - Added home.lan. to search paths
[main] DEBUG org.xbill.DNS.config.ResolvConfResolverConfigProvider - Added /192.168.0.250:53 to nameservers
[main] DEBUG org.xbill.DNS.config.ResolvConfResolverConfigProvider - Added /8.8.8.8:53 to nameservers
[main] DEBUG org.xbill.DNS.Lookup - Lookup for google.com./A, cache answer: unknown
[main] DEBUG org.xbill.DNS.ExtendedResolver - Sending google.com./A, id=59702 to resolver 0 (SimpleResolver [/192.168.0.250:53]), attempt 1 of 3
[main] TRACE org.xbill.DNS.Compression - Looking for google.com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding google.com. at 12
[main] TRACE org.xbill.DNS.Compression - Looking for com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding com. at 19
[main] DEBUG org.xbill.DNS.SimpleResolver - Sending google.com./A, id=59702 to udp/192.168.0.250:53
[main] TRACE org.xbill.DNS.SimpleResolver - Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 59702
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	google.com., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 39 bytes
[main] DEBUG org.xbill.DNS.Client - Starting dnsjava NIO selector thread
[dnsjava NIO selector] TRACE org.xbill.DNS.Client - 39b (UDP write):	E9 36 01 00 00 01 00 00 00 00 00 01 06 67 6F 6F 67 6C 
			65 03 63 6F 6D 00 00 01 00 01 00 00 29 05 00 00 00 00 
			00 00 00 

[dnsjava NIO selector] TRACE org.xbill.DNS.Client - 55b (UDP read):	E9 36 81 80 00 01 00 01 00 00 00 01 06 67 6F 6F 67 6C 65 03 63 
		6F 6D 00 00 01 00 01 C0 0C 00 01 00 01 00 00 00 76 00 04 8E FA 
		BB EE 00 00 29 02 00 00 00 00 00 00 00 

[ForkJoinPool.commonPool-worker-1] TRACE org.xbill.DNS.Name - currently 30, pointer to 12
[ForkJoinPool.commonPool-worker-1] TRACE org.xbill.DNS.Name - current name '@', seeking to 12
[main] DEBUG org.xbill.DNS.Cache - Caching successful for google.com./A
[main] DEBUG org.xbill.DNS.Lookup - Queried google.com./A, id=59702: successful
List(google.com.		118	IN	A	142.250.187.238)
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com./A, cache answer: unknown
[main] DEBUG org.xbill.DNS.ExtendedResolver - Sending facebook.com./A, id=53295 to resolver 0 (SimpleResolver [/192.168.0.250:53]), attempt 1 of 3
[main] TRACE org.xbill.DNS.Compression - Looking for facebook.com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding facebook.com. at 12
[main] TRACE org.xbill.DNS.Compression - Looking for com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding com. at 21
[main] DEBUG org.xbill.DNS.SimpleResolver - Sending facebook.com./A, id=53295 to udp/192.168.0.250:53
[main] TRACE org.xbill.DNS.SimpleResolver - Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 53295
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	facebook.com., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 41 bytes
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com./A, id=53295 failed using resolver ExtendedResolver of [SimpleResolver [/192.168.0.250:53], SimpleResolver [/8.8.8.8:53]]
java.net.SocketTimeoutException
	at org.xbill.DNS.Resolver.send(Resolver.java:157)
	at org.xbill.DNS.Lookup.lookup(Lookup.java:524)
	at org.xbill.DNS.Lookup.resolve(Lookup.java:581)
	at org.xbill.DNS.Lookup.run(Lookup.java:599)
	at common.dns$.main(dns.scala:22)
	at common.dns.main(dns.scala)
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com.home.lan./A, cache answer: unknown
[main] DEBUG org.xbill.DNS.ExtendedResolver - Sending facebook.com.home.lan./A, id=25160 to resolver 0 (SimpleResolver [/192.168.0.250:53]), attempt 1 of 3
[main] TRACE org.xbill.DNS.Compression - Looking for facebook.com.home.lan., found -1
[main] TRACE org.xbill.DNS.Compression - Adding facebook.com.home.lan. at 12
[main] TRACE org.xbill.DNS.Compression - Looking for com.home.lan., found -1
[main] TRACE org.xbill.DNS.Compression - Adding com.home.lan. at 21
[main] TRACE org.xbill.DNS.Compression - Looking for home.lan., found -1
[main] TRACE org.xbill.DNS.Compression - Adding home.lan. at 25
[main] TRACE org.xbill.DNS.Compression - Looking for lan., found -1
[main] TRACE org.xbill.DNS.Compression - Adding lan. at 30
[main] DEBUG org.xbill.DNS.SimpleResolver - Sending facebook.com.home.lan./A, id=25160 to udp/192.168.0.250:53
[main] TRACE org.xbill.DNS.SimpleResolver - Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25160
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	facebook.com.home.lan., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 50 bytes
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com.home.lan./A, id=25160 failed using resolver ExtendedResolver of [SimpleResolver [/192.168.0.250:53], SimpleResolver [/8.8.8.8:53]]
java.net.SocketTimeoutException
	at org.xbill.DNS.Resolver.send(Resolver.java:157)
	at org.xbill.DNS.Lookup.lookup(Lookup.java:524)
	at org.xbill.DNS.Lookup.resolve(Lookup.java:581)
	at org.xbill.DNS.Lookup.run(Lookup.java:606)
	at common.dns$.main(dns.scala:22)
	at common.dns.main(dns.scala)
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com./A, cache answer: unknown
[main] DEBUG org.xbill.DNS.ExtendedResolver - Sending facebook.com./A, id=1947 to resolver 0 (SimpleResolver [/192.168.0.250:53]), attempt 1 of 3
[main] TRACE org.xbill.DNS.Compression - Looking for facebook.com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding facebook.com. at 12
[main] TRACE org.xbill.DNS.Compression - Looking for com., found -1
[main] TRACE org.xbill.DNS.Compression - Adding com. at 21
[main] DEBUG org.xbill.DNS.SimpleResolver - Sending facebook.com./A, id=1947 to udp/192.168.0.250:53
[main] TRACE org.xbill.DNS.SimpleResolver - Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1947
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	facebook.com., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 41 bytes
[main] DEBUG org.xbill.DNS.Lookup - Lookup for facebook.com./A, id=1947 failed using resolver ExtendedResolver of [SimpleResolver [/192.168.0.250:53], SimpleResolver [/8.8.8.8:53]]
java.net.SocketTimeoutException
	at org.xbill.DNS.Resolver.send(Resolver.java:157)
	at org.xbill.DNS.Lookup.lookup(Lookup.java:524)
	at org.xbill.DNS.Lookup.resolve(Lookup.java:581)
	at org.xbill.DNS.Lookup.run(Lookup.java:614)
	at common.dns$.main(dns.scala:22)
	at common.dns.main(dns.scala)
java.lang.NullPointerException
	at common.dns$.main(dns.scala:26)
	at common.dns.main(dns.scala)
[dnsjava NIO selector] DEBUG org.xbill.DNS.Client - dnsjava NIO selector thread stopped
[ForkJoinPool.commonPool-worker-3] DEBUG org.xbill.DNS.ExtendedResolver - Failed to resolve facebook.com./A, id=53295 with resolver 0 (SimpleResolver [/192.168.0.250:53]) on attempt 1 of 3, reason=java.io.EOFException: Client is closing
[ForkJoinPool.commonPool-worker-4] DEBUG org.xbill.DNS.ExtendedResolver - Failed to resolve facebook.com.home.lan./A, id=25160 with resolver 0 (SimpleResolver [/192.168.0.250:53]) on attempt 1 of 3, reason=java.io.EOFException: Client is closing
[ForkJoinPool.commonPool-worker-5] DEBUG org.xbill.DNS.ExtendedResolver - Failed to resolve facebook.com./A, id=1947 with resolver 0 (SimpleResolver [/192.168.0.250:53]) on attempt 1 of 3, reason=java.io.EOFException: Client is closing
timed out
```

**Comment by ibauersachs** [comment]
Sorry, missed your answer. I'm not sure what's going in this particular setup, however there are a couple of things that interact here:
- two DNS servers are detected, 192.168.0.250 and 8.8.8.8
- a search path of `home.lan` is used
- 192.168.0.250 seems unresponsive, at least sometimes

If `Lookup` encounters an exception during resolving, it tries to go on. This happens e.g. when resolving `facebook.com` fails (a relative name, no dot at the end) and tries again as `facebook.com.home.lan.` (now an absolute name). Both resolutions fail, and the end result is a fail. Now, `ExtendedResolver` has a timeout, 10s by default. But each `SimpleResolver` used by the ExtendedResolver _also_ has a timeout, 5s by default.
In your case with two resolvers, this should line up, i.e. when 192.168.0.250 fails, it should fall back to 8.8.8.8 within the timeout and give you a result (my local tests behave exactly like that). Somehow that doesn't happen and I have no idea why. There might be a hint if you could add a file named `simplelogger.properties` to your classpath with the following content to show the timings of the debug log and post it again:
```
org.slf4j.simpleLogger.dateTimeFormat=yyyy-MM-dd HH:mm:ss:SSSZ
org.slf4j.simpleLogger.showDateTime=true
```

Some general recommendations:
- use absolute domain names if you can,
- alternatively, set an explicit search path if you can
- use reliable nameservers, 192.168.0.250 is probably your home router doing some DNS proxying and routers are known for behaving badly

**Comment by eberleant** [comment]
By the way, I'm having the same or a similar issue on MacOS Monterey (12.4). Running Lookup.run() twice always seems to work the first time, but time out the second time. The same code works on my other laptop, which is on MacOS Catalina (10.15.7). I'm using dnsjava version 3.0.0 but also tried on 3.5.1, with the same result. A colleague (also using Monterey) got the same result on a different network.

Let me know if I should open a separate ticket for this.

Code:
```
var lookup = new Lookup("example.com", Type.A);
lookup.run();
Lookup.refreshDefault(); // reset cache
var lookup = new Lookup("example.com", Type.A);
lookup.run();
```

Trace logs:
```
[2022-07-29T17:19:04.011-0400] [DEBUG] [.xbill.DNS.config.ResolvConfResolverConfigProvider] [main                               ]: Added /192.168.0.1:53 to nameservers
[2022-07-29T17:19:04.055-0400] [DEBUG] [figProvider$InnerJndiContextResolverConfigProvider] [main                               ]: Added /192.168.0.1:53 to nameservers
[2022-07-29T17:19:04.077-0400] [DEBUG] [org.xbill.DNS.Lookup                              ] [main                               ]: lookup example.com. A, cache answer: unknown
[2022-07-29T17:19:04.114-0400] [DEBUG] [org.xbill.DNS.ExtendedResolver                    ] [main                               ]: Sending example.com./A, id=27083 to resolver 0 (SimpleResolver [/192.168.0.1:53]), attempt 1 of 3
[2022-07-29T17:19:04.116-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Looking for example.com., found -1
[2022-07-29T17:19:04.117-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Adding example.com. at 12
[2022-07-29T17:19:04.117-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Looking for com., found -1
[2022-07-29T17:19:04.118-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Adding com. at 20
[2022-07-29T17:19:04.118-0400] [DEBUG] [org.xbill.DNS.SimpleResolver                      ] [main                               ]: Sending example.com./A, id=27083 to udp/192.168.0.1:53
[2022-07-29T17:19:04.119-0400] [TRACE] [org.xbill.DNS.SimpleResolver                      ] [main                               ]: Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27083
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	example.com., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 40 bytes
[2022-07-29T17:19:04.123-0400] [DEBUG] [org.xbill.DNS.Client                              ] [main                               ]: Starting dnsjava NIO selector thread
[2022-07-29T17:19:04.169-0400] [TRACE] [org.xbill.DNS.Client                              ] [dnsjava NIO selector               ]: 40b (UDP write):	69 CB 01 00 00 01 00 00 00 00 00 01 07 65 78 61 6D 70 
			6C 65 03 63 6F 6D 00 00 01 00 01 00 00 29 05 00 00 00 
			00 00 00 00 

[2022-07-29T17:19:04.190-0400] [TRACE] [org.xbill.DNS.Client                              ] [dnsjava NIO selector               ]: 56b (UDP read):	69 CB 81 80 00 01 00 01 00 00 00 01 07 65 78 61 6D 70 6C 65 03 
		63 6F 6D 00 00 01 00 01 C0 0C 00 01 00 01 00 00 4B 0A 00 04 5D 
		B8 D8 22 00 00 29 02 00 00 00 00 00 00 00 

[2022-07-29T17:19:04.194-0400] [TRACE] [org.xbill.DNS.Name                                ] [ForkJoinPool.commonPool-worker-3   ]: currently 31, pointer to 12
[2022-07-29T17:19:04.195-0400] [TRACE] [org.xbill.DNS.Name                                ] [ForkJoinPool.commonPool-worker-3   ]: current name '@', seeking to 12
[2022-07-29T17:19:04.198-0400] [DEBUG] [org.xbill.DNS.Cache                               ] [main                               ]: caching successful for example.com.
[2022-07-29T17:19:04.199-0400] [DEBUG] [org.xbill.DNS.Lookup                              ] [main                               ]: Queried example.com. A: successful
[2022-07-29T17:19:04.199-0400] [DEBUG] [org.xbill.DNS.Lookup                              ] [main                               ]: lookup example.com. A, cache answer: unknown
[2022-07-29T17:19:04.200-0400] [DEBUG] [org.xbill.DNS.ExtendedResolver                    ] [main                               ]: Sending example.com./A, id=60079 to resolver 0 (SimpleResolver [/192.168.0.1:53]), attempt 1 of 3
[2022-07-29T17:19:04.201-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Looking for example.com., found -1
[2022-07-29T17:19:04.201-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Adding example.com. at 12
[2022-07-29T17:19:04.202-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Looking for com., found -1
[2022-07-29T17:19:04.202-0400] [TRACE] [org.xbill.DNS.Compression                         ] [main                               ]: Adding com. at 20
[2022-07-29T17:19:04.203-0400] [DEBUG] [org.xbill.DNS.SimpleResolver                      ] [main                               ]: Sending example.com./A, id=60079 to udp/192.168.0.1:53
[2022-07-29T17:19:04.203-0400] [TRACE] [org.xbill.DNS.SimpleResolver                      ] [main                               ]: Query:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 60079
;; flags: rd ; qd: 1 an: 0 au: 0 ad: 1 
;; QUESTIONS:
;;	example.com., type = A, class = IN

;; ANSWERS:

;; AUTHORITY RECORDS:

;; ADDITIONAL RECORDS:
.			0	CLASS1280	OPT	 ; payload 1280, xrcode 0, version 0, flags 0

;; Message size: 40 bytes
[2022-07-29T17:19:14.279-0400] [DEBUG] [org.xbill.DNS.Lookup                              ] [main                               ]: Lookup failed using resolver ExtendedResolver of [SimpleResolver [/192.168.0.1:53]]
java.net.SocketTimeoutException: null
	at org.xbill.DNS.Resolver.send(Resolver.java:156) ~[dnsjava-3.0.0.jar:?]
	at org.xbill.DNS.Lookup.lookup(Lookup.java:493) [dnsjava-3.0.0.jar:?]
	at org.xbill.DNS.Lookup.resolve(Lookup.java:543) [dnsjava-3.0.0.jar:?]
	at org.xbill.DNS.Lookup.run(Lookup.java:561) [dnsjava-3.0.0.jar:?]
	at org.prss.metapub.RadioDnsRegistrationChecker.printRecords(RadioDnsRegistrationChecker.java:60) [metapub-2022.5.jar:?]
	at org.prss.metapub.RadioDnsRegistrationChecker.main(RadioDnsRegistrationChecker.java:44) [metapub-2022.5.jar:?]
```

**Comment by ibauersachs** [comment]
@eberleant 
Please use version 3.5.1 only, there were so many fixes since 3.0 that debugging an old version is pointless.

Regardless, in your log is something that I missed previously: the second query isn't even sent/written to the network. Can you please take a thread dump _after_ the first query and _before_ the second one times out?

For simplicity, let's take the `ExtendedResolver` out of the picture, and use a single resolver like this (don't call `Lookup.refreshDefault();` or the resolver is reset too, use l.setCache(null) instead):

```java
Resolver r = new SimpleResolver("192.168.10.1");
Lookup l = new Lookup("example.com.", Type.A);
l.setResolver(r);
l.setCache(null);
for (int i = 0; i < 3; i++) {
  Record[] results = l.run();
  if (results != null) {
    System.out.println(Arrays.asList(results));
  } else {
    System.out.println(l.getErrorString());
  }
}
```

**Comment by eberleant** [comment]
Thanks for the response! So, I updated to 3.5.1 and it works perfectly- I must've done something silly like forgetting to rebuild when I tested the newer version before. Sorry for wasting your time! and thanks for a great library.
